### PR TITLE
urdfdom: add recipe

### LINKED ISSUE #17345: [request] urdfdom
### Package Name/Version

urdfdom

### Webpage

https://github.com/ros/urdfdom

### Source code

https://github.com/ros/urdfdom

### Description of the library/tool

The urdfdom library provides core data structures and a simple XML parsers for populating the class data structures from an [URDF (U-Robot Description Format) file](https://www.mathworks.com/help/sm/ug/urdf-model-import.html).

A core ROS (Robot Operating System) library for dealing with URDF files.

[Packaging status](https://repology.org/project/urdfdom/versions)

### PR DESCRIPTION
**Pull request by valgur**
Specify library name and version:  **urdfdom/3.1.1**

Resolves #17345.
Requires
* #17346
* #18779

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by valgur** [comment]
```
ERROR: Missing prebuilt package for 'tinyxml/2.6.2'
Check the available packages using 'conan list tinyxml/2.6.2:* -r=remote'
or try to build locally from sources using the '--build=tinyxml/2.6.2' argument
```
Not sure how to handle this. @uilianries, @SSE4 or anyone else - do you have any suggestions?

**Comment by uilianries** [comment]
@valgur The package `tinyxml` is only distributed with the option `with_tls=False`, which means, there is no pre-built package available and is not possible to generate custom configurations in ConanCenterIndex CI, neither distribute them by Conan Center.

So, is it possible to build this project with `tinyxml:with_tls=False`? Otherwise, it will not not be possible to distribute pre-built packages.

**Comment by valgur** [comment]
Thanks!
Unfortunately, urdfdom will fail to build if `tinyxml:with_stl=True` is not set.
Does this effectively mean non-default dependency options are not allowed on ConanCenter and this recipe cannot be added?

**Comment by uilianries** [comment]
> Does this effectively mean non-default dependency options are not allowed on ConanCenter and this recipe cannot be added?

You still can add the recipe, be reviewed and eventually merged, but it will not generate packages as non-default configuration are not allowed. Thus, Conan Center will provide the recipe only.

**Comment by valgur** [comment]
Hmm, would changing the tinyxml default to `with_stl=True` be a valid option, maybe? Adding interfaces for STL to the default build does not seem too unreasonable.

**Review by memsharded — COMMENTED** [review]
Looking good, thanks very much for your contribution.
Just one minor comment (and we have to discuss what can be down with the tinyxml tls thing, this is unfortunate...)

File: recipes/urdfdom/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    def configure(self):
+        if self.options.shared:
+            self.options.rm_safe("fPIC")
+        self.options["tinyxml"].with_stl = True
+
+    def layout(self):
+        cmake_layout(self, src_folder="src")
+
+    def requirements(self):
+        self.requires("tinyxml/2.6.2", transitive_headers=True, transitive_libs=True)
```
[the remark this answers is not in the document]
Comment (valgur) [reply recipes/urdfdom/all/conanfile.py]: Fixed. Thanks!

[the remark this answers is not in the document]
Comment (valgur) [reply recipes/urdfdom/all/conanfile.py]: Also, as a side note, I think the [docs for `transitive_libs`](https://docs.conan.io/2/reference/conanfile/methods/requirements.html#transitive-libs) would benefit from explicitly mentioning something like "`transitive_libs` is set to the value of `libs` by default". Of course, it makes sense for that to be the case if you think about it for more than two seconds, but I did not feel certain about that until re-watching the Advanced Dependencies Model in Conan 2.0 presentation again now.

**Comment by uilianries** [comment]
I started an internal CI build to generate tinyxml. The result should be available at https://c3i.jfrog.io/c3i/misc-v2/summary.html?json=https://c3i.jfrog.io/c3i/misc-v2/logs/prod-v2/TapaholesRepo/57/summary.json as soon as ready.

**Comment by RubenRBS** [comment]
https://github.com/conan-io/conan-center-index/pull/18779 is still open, we should focus on deciding what to do with that beforehand :)

**Comment by RubenRBS** [comment]
tinyxml has now been fixed and the remaining errors are real linter ones :)

**Review by RubenRBS — DISMISSED** [review]

File: recipes/urdfdom/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        export_conandata_patches(self)
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
+        if self.options.shared:
+            self.options.rm_safe("fPIC")
+        self.options["tinyxml"].with_stl = True
```
Comment (RubenRBS) [inline recipes/urdfdom/all/conanfile.py:43]:
```suggestion
```

No need for this line now that tinyxml default's to it

**Review by RubenRBS — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Pull request merged**